smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptJmesPathVisitor.java

```diff
@@ -0,0 +1,334 @@
+        executionContext += ".";
+        executionContext += expression.getName();
+        return null;
+    }
+
+    @Override
+    public Void visitIndex(IndexExpression expression) {
+        if (expression.getIndex() >= 0) {
+            executionContext += ("[" + expression.getIndex() + "]");
+        } else {
+            executionContext += "[" + executionContext + ".length";
+            executionContext += " - " + Math.abs(expression.getIndex()) + "]";
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitLiteral(LiteralExpression expression) {
+        switch (expression.getType()) {
+            case STRING:
+                executionContext = "\"" + expression.getValue().toString() + "\"";
+                break;
+            default:
+                // All other options are already valid js literials.
+                // (BOOLEAN, ANY, ARRAY, NULL, NUMBER, OBJECT, EXPRESSION)
+                executionContext = expression.getValue().toString();
+                break;
+        }
+        return null;
+    }
+
+    @Override
+    public Void visitMultiSelectList(MultiSelectListExpression expression) {
+        ArrayList<String> evaluators = new ArrayList<String>();
+
+        String executionContextInital = executionContext;
+
+        expression.getExpressions().forEach((JmespathExpression exp) -> {
+            exp.accept(this);
+            evaluators.add(executionContext);
+            executionContext = executionContextInital;
+        });
+
+        String resultScope = makeNewScope("result_");
+        writer.write("let $L = [];", resultScope);
+        for (String evaluator : evaluators) {
+            writer.write("$L.push($L);", resultScope, evaluator);
+        }
+        writer.write("$L = $L;", executionContext, resultScope);
+
+        return null;
+    }
+
+    @Override
+    public Void visitMultiSelectHash(MultiSelectHashExpression expression) {
+        throw new CodegenException("TypeScriptJmesPath visitor not implemented MultiSelectHashExpression");
+    }
+
+    @Override
+    public Void visitAnd(AndExpression expression) {
+        String initialContext = executionContext;
+
+        expression.getLeft().accept(this);
+        String leftContext = executionContext;
+        executionContext = initialContext;
+
+        expression.getRight().accept(this);
+        String rightContext = executionContext;
+
+        executionContext = String.format("(%s && %s)", leftContext, rightContext);
+        return null;
+    }
+
+    @Override
+    public Void visitOr(OrExpression expression) {
+        String initialContext = executionContext;
+
+        expression.getLeft().accept(this);
+        String leftContext = executionContext;
+        executionContext = initialContext;
+
+        expression.getRight().accept(this);
+        String rightContext = executionContext;
+
+        executionContext = String.format("(%s || %s)", leftContext, rightContext);
```

Comment:
I don't think the JavaScript and / or semantics match the truth table for JMESPath (cf https://jmespath.org/specification.html#or-expressions)

Reply:
You're right, we discussed offline. To match smithy's truth table in JS we need to do something like:
```
((A || B) && (B || A))
```